lib/gambit/list/list.scm

```diff
@@ -607,22 +607,26 @@
                 (macro-check-list x 2 (assoc obj lst c)
                   #f))))))))
 
-(define-prim (##remq elem lst)
-
-  (include "~~lib/gambit/prim/prim#.scm") ;; map fx+ to ##fx+, etc
-
-  (define (rem elem lst)
-    (if (pair? lst)
-        (let* ((lst2 (rem elem (cdr lst)))
-               (x (car lst)))
-          (if (eq? x elem)
-              lst2
-              (if (eq? lst2 (cdr lst))
-                  lst
-                  (cons x lst2))))
-        '()))
-
-  (rem elem lst))
+(define-macro (##filter-macro lst var test-expr)
+  `(let recurse ((lst ,lst))
+     (if (pair? lst)
+         (let* ((lst2 (recurse (cdr lst)))
+                (,var (car lst)))
+           (if ,test-expr
+               (if (eq? lst2 (cdr lst))
```

Comment:
This is not quite right in a multithreaded Scheme because the `cdr` of `lst` could have been modified by a second thread to make the test `(eq? lst2 (cdr lst))` succeed (it is very unlikely, but still possible).  It is better to do `(cdr lst)` once and save the value in a `rest` variable in the `let*`, and then test `(eq? lst2 rest)`.